src/index.ts

```diff
@@ -0,0 +1,316 @@
+    this.field = field;
+    this.sessionData = data;
+  }
+
+  public next(values: Values): [Tracking, undefined] | [undefined, Error] {
+    if (this.queryID === "") {
+      this.queryID = newQueryID();
+      this.sequence = 0;
+    } else {
+      this.sequence++;
+    }
+
+    return [
+      {
+        type: this.trackingType,
+        query_id: this.queryID,
+        sequence: this.sequence,
+        field: this.field,
+        data: this.sessionData
+      } as Tracking,
+      undefined
+    ];
+  }
+
+  public reset(): void {
+    this.queryID = "";
+    this.sequence = 0;
+  }
+}
+
+export type ClientOption = (client: Client) => void;
+
+/**
+ * withEndpoint constructs a [[ClientOption]] that modifies the endpoint used by the client.
+ */
+export const withEndpoint = (endpoint: string) => (client: Client) => {
+  client.endpoint = endpoint;
+};
+
+/**
+ * Client takes configuration and constructs pipelines clients.
+ *
+ * ```javascript
+ * const client = new Client("<project>", "<collection>");
+ * const webPipeline = client.pipeline("website");
+ * // webPipeline.search(...);
+ * ```
+ *
+ * An optional array of [[ClientOption]] may be given to the client constructor to modify it's behaviour.
+ *
+ * ```javascript
+ * const client = new Client("<project>", "<collection>", withEndpoint("https://example.com"));
+ * ```
+ */
+export class Client {
+  public project: string;
+  public collection: string;
+  public endpoint: string;
+
+  public constructor(
+    project: string,
+    collection: string,
+    opts: ClientOption[] = []
+  ) {
+    this.project = project;
+    this.collection = collection;
+    this.endpoint = "https://jsonapi.sajari.net";
+    opts.forEach(opt => {
+      opt(this);
+    });
+  }
+
+  /**
+   * pipeline returns a new [[Pipeline]].
+   */
+  public pipeline(name: string): Pipeline {
+    return new PipelineImpl(this, name);
+  }
+}
+
+export interface Key {
+  field: string;
+  value: any;
+}
+
+export type SearchCallback = (
+  error: SearchError | undefined,
+  results: Results | undefined,
+  values: Values | undefined
+) => void;
+
+export type AddCallback = (key: Key, error: SearchError) => void;
+
+export interface SJRecord {
+  [id: string]: any;
+}
+
+/**
+ * PipelineImpl is private to prevent users constructing it themselves.
+ */
+class PipelineImpl {
+  public client: Client;
+  public name: string;
+  public endpoint: string = "sajari.api.pipeline.v1.Query/Search";
+
+  public constructor(client: Client, name: string) {
+    this.client = client;
+    this.name = name;
+  }
+
+  public search(
+    values: Values,
+    session: Session,
+    callback: SearchCallback
+  ): void {
+    const [tracking, error] = session.next(values);
+    if (error) {
+      callback(
+        newError("error getting next session: " + error),
+        undefined,
+        undefined
+      );
+      return;
+    }
+
+    const requestBody = JSON.stringify({
+      // tslint:disable:object-literal-sort-keys
```

Comment:
?

Reply:
This was me, stops tslint for requiring everything to alphabetically sorted which didn't make a lot of sense in this case.